src/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs

```diff
@@ -0,0 +1,412 @@
+        /// <returns></returns>
+        public T WithOptions(ApplicationOptions applicationOptions)
+        {
+            WithClientId(applicationOptions.ClientId);
+            WithRedirectUri(applicationOptions.RedirectUri);
+            WithTenant(applicationOptions.Tenant);
+            WithLoggingLevel(applicationOptions.LogLevel);
+            WithComponent(applicationOptions.Component);
+            WithEnablePiiLogging(applicationOptions.EnablePiiLogging);
+            WithDefaultPlatformLoggingEnabled(applicationOptions.IsDefaultPlatformLoggingEnabled);
+            WithSliceParameters(applicationOptions.SliceParameters);
+
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="component"></param>
+        /// <returns></returns>
+        public T WithComponent(string component)
+        {
+            Config.Component = GetValueIfNotEmpty(Config.Component, component);
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="sliceParameters"></param>
+        /// <returns></returns>
+        public T WithSliceParameters(string sliceParameters)
+        {
+            Config.SliceParameters = GetValueIfNotEmpty(Config.SliceParameters, sliceParameters);
+            return (T)this;
+        }
+
+        internal ApplicationConfiguration BuildConfiguration()
+        {
+            // todo: final validation/sanity checks here...
+
+            // validate that we only have ONE default authority
+            if (Config.Authorities.Where(x => x.IsDefault).ToList().Count > 1)
+            {
+                throw new InvalidOperationException("More than one default authority was configured.");
+            }
+
+            if (!Config.Authorities.Any())
+            {
+                throw new InvalidOperationException("No authorities were configured.");
+            }
+
+            return Config;
+        }
+
+        /// <summary>
+        ///     Note, we'll probably want authority improvements, but bootstrapping with this...
+        /// </summary>
+        /// <param name="authorityUri"></param>
+        /// <param name="validateAuthority"></param>
+        /// <param name="isDefaultAuthority"></param>
+        /// <returns></returns>
+        public T WithAuthority(string authorityUri, bool validateAuthority, bool isDefaultAuthority)
+        {
+            Config.AddAuthorityInfo(AuthorityInfo.FromAuthorityUri(authorityUri, validateAuthority, isDefaultAuthority));
+            return (T)this;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="cloudUrl"></param>
+        /// <param name="tenantId"></param>
+        /// <param name="isDefaultAuthority"></param>
+        /// <returns></returns>
+        public T WithAadAuthority(
+            string cloudUrl,
+            Guid tenantId,
+            bool isDefaultAuthority)
```

Comment:
Do we need the concept of DefaultAuthority?  This came from the Android API review.  There, they add multiple "known authorities" in their PCA configuration and then I believe they have to send in the specific authority when they call AT/ATS.  I added this so that we can have the developer send in multiple known authorities and set one as the "default" which we will use when actually making the calls.  But all of MSAL (thus far) assumes that a PCA has one authority (except in a few method variations where we can override it).  

This is something we need to finalize in the API surface.